### PR TITLE
Implement HttpTelemetry for HTTP/3

### LINKED ISSUE #40896: HttpTelemetry instrumentation for HTTP/3
Newly added `System.Net.Http` [Telemetry](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs) should be expanded to HTTP/3 in the future. 

Explicitly disabled in #40839.

### PR DESCRIPTION
**Pull request by MihaZupan**
Fixes #40896

### CODE REVIEW

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs [hunk 1]
```diff
@@ -148,41 +159,64 @@ private void CheckForShutdown()
...
-                    await waitTask.ConfigureAwait(false);
+                        // Wait for an available stream (based on QUIC MAX_STREAMS) if there isn't one available yet.
+                        await waitTask.ConfigureAwait(false);
+                    }
+                }
+                finally
+                {
+                    if (HttpTelemetry.Log.IsEnabled() && queueDuration.IsActive)
+                    {
+                        HttpTelemetry.Log.Http30RequestLeftQueue(queueDuration.GetElapsedTime().TotalMilliseconds);
```
Comment (ManickaP) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs:218]: In case the connection gets closed in the meantime, the request will get retried on a different connection (line 224). I'm not sure what counts into the "waiting time in the queue" in such case, but this will report as request left the queue.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs [hunk 2]
```diff
@@ -283,7 +283,7 @@ public async Task<HttpResponseMessage> SendAsync(CancellationToken cancellationT
                 if (cancellationToken.IsCancellationRequested)
                 {
                     _stream.AbortWrite((long)Http3ErrorCode.RequestCancelled);
-                    throw new OperationCanceledException(ex.Message, ex, cancellationToken);
+                    throw new TaskCanceledException(ex.Message, ex, cancellationToken);
```
Comment (ManickaP) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:286]: Why this change? Is it to unify exception types with H/2, H/1.1?

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs [hunk 3]
```diff
@@ -902,7 +902,7 @@ private async ValueTask<Http3Connection> GetHttp3ConnectionAsync(HttpRequestMess
         [SupportedOSPlatform("windows")]
         [SupportedOSPlatform("linux")]
         [SupportedOSPlatform("macos")]
-        private async ValueTask<HttpResponseMessage?> TrySendUsingHttp3Async(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
+        private async ValueTask<HttpResponseMessage?> TrySendUsingHttp3Async(HttpRequestMessage request, CancellationToken cancellationToken)
```
Comment (ManickaP) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:905]: Thanks for the `async` removal!

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs [hunk 4]
```diff
@@ -343,6 +345,8 @@ private async Task ReadResponseAsync(CancellationToken cancellationToken)
             while ((int)_response.StatusCode < 200);
 
             _headerState = HeaderState.TrailingHeaders;
+
+            if (HttpTelemetry.Log.IsEnabled()) HttpTelemetry.Log.ResponseHeadersStop();
```
Comment (ManickaP) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:349]: This is not exactly the moment when the headers get sent, but it gets muddy with the gathered buffer with content. How does this compare with H/2 and H/1.1? Is is worth trying to refine the moment or not?

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs (hunk 2 above)
Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:286]: Right, this is something that both 1.1 and 2 seem to be doing.
E.g. for 1.1:
https://github.com/dotnet/runtime/blob/dc9979136cf0ff1addbc1a4c0f06ea4b65f002d4/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L842

It seems harmless enough of a change (given TCE derives from OCE), or we could update the [failure tests](https://github.com/dotnet/runtime/blob/b3847f42ae95cc082a6b28575d4d847187b31a7b/src/libraries/System.Net.Http/tests/FunctionalTests/TelemetryTest.cs#L238) to expect OCE instead.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs (hunk 1 above)
Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs:218]: This should be rare-ish to happen, but yes, you may see multiple `RequestLeftQueue` events for a given request.

Similarly with 1.1, there is a race where we would retry if the server disconnects [here](https://github.com/dotnet/runtime/blob/dc9979136cf0ff1addbc1a4c0f06ea4b65f002d4/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L587-L593), logging `RequestLeftQueue` twice.
Similarly with resets on HTTP/2, we may re-insert the request to the queue.

I'm not sure we can/should improve that somehow. I feel it's still valuable to include the time on queue here even if the request will just end up back in the queue.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs (hunk 4 above)
Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:349]: On HTTP/2, we will actually include the time to write out the headers on the wire ([code](https://github.com/dotnet/runtime/blob/dc9979136cf0ff1addbc1a4c0f06ea4b65f002d4/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1529-L1587)). That is because the time when the headers are sent is well-defined since we immediately serialize the action on the multiplexed connection.

On HTTP/1.1 and 3, however, the headers may not end up being sent right away, but only with the first flush of the request content, which may be under user's control.
We could report a more accurate timestamp of when the headers are actually sent, but these events being Start/Stop complicates the matter. If ThreadPool activity tracking is on, I don't see how we would get the async scope for Start/Stop to match, given Stop would result as part of the content copying outside of `SendAsync`.

Comment (ManickaP) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:349]: I guess this is good enough and not worth the trouble. Especially, if we're already not pedantic about the exact times in H/1.1

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs (hunk 1 above)
Comment (ManickaP) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs:218]: If we do similar thing with H/1.1 and H/2 then this is fine. I'm mostly trying to make sure we keep similar behavior between the protocols so that one counter doesn't mean something else based on HTTP version.

**Review by ManickaP — APPROVED** [review]
Thanks!
